Implement caching for Babel package builds

b-ryu opened the pull request:
This PR adds caching to `createCompileBabelStep` in `plugin-javascript` (used for transpiling packages). Currently TS builds are cached, however when building JS using the Babel CLI we rebuild each package from scratch (Babel CLI doesn't seem to support caching OOTB).

How the cache works is that it generates a hash based on several compilation-relevant factors (including Babel configs, deps, etc.) in addition to the latest modified time (`mtimeMs`) of the files in the package's `src` folder. If it doesn't match the hash stored in `.sewing-kit/cache/babel` (i.e. the config/deps have changed, or the files have been modified since the last build), then it rebuilds. Otherwise it skips. This helps reduce warm build times in workspaces with many packages but with usually only a few being worked on at any given time.

### To do
- [x] Add tests

### Benchmarking

_Note: Quilt build times vary somewhat, particularly for cold builds. The "Before" ones listed below are on the fast end, wasn't able to get a good "average" time 😕_

#### In Sewing Kit (self-build)

| | Before | After |
| --- | --- | --- |
| **Cold** | <img width="316" alt="Screen Shot 2020-08-17 at 11 19 48 AM" src="https://user-images.githubusercontent.com/46533681/90412852-a27e3200-e07b-11ea-9e48-ca2af9f2da0d.png"> | <img width="226" alt="Screen Shot 2020-08-17 at 11 15 52 AM" src="https://user-images.githubusercontent.com/46533681/90412456-208e0900-e07b-11ea-8fa1-c1c7aa659eaf.png"> |
| **Warm** | <img width="350" alt="Screen Shot 2020-08-17 at 11 20 08 AM" src="https://user-images.githubusercontent.com/46533681/90412875-aad66d00-e07b-11ea-971f-3dfb52296a76.png"> | <img width="278" alt="Screen Shot 2020-08-17 at 11 16 24 AM" src="https://user-images.githubusercontent.com/46533681/90412482-27b51700-e07b-11ea-8e1d-e6141857a0fa.png"> | 

#### In [Quilt](https://github.com/Shopify/quilt/tree/1522-sk-next)

| | Before | After |
| --- | --- | --- |
| **Cold** | <img width="253" alt="Screen Shot 2020-08-17 at 12 07 23 PM" src="https://user-images.githubusercontent.com/46533681/90423219-1293b480-e08a-11ea-8f97-5cf5b5a53927.png"> | <img width="252" alt="Screen Shot 2020-08-17 at 1 14 24 PM" src="https://user-images.githubusercontent.com/46533681/90424146-a2862e00-e08b-11ea-8ec4-a1b67320a474.png"> |
| **Warm** | <img width="249" alt="Screen Shot 2020-08-17 at 12 10 58 PM" src="https://user-images.githubusercontent.com/46533681/90423238-19222c00-e08a-11ea-8385-fa4584052016.png"> | <img width="243" alt="Screen Shot 2020-08-17 at 1 15 16 PM" src="https://user-images.githubusercontent.com/46533681/90424197-b598fe00-e08b-11ea-86b9-9fab107f3f9f.png"> |

Commit:
Implement caching for Babel builds

Commit:
Refactor

Commit:
Use more params in hash

Commit:
Refactor

ismail-syed commented on packages/plugin-javascript/src/utilities.ts:
```diff
@@ -1,5 +1,8 @@
+import {sync as glob} from 'glob';
+import {statSync as stat} from 'fs';
 import {createHash} from 'crypto';
+import {readFile, writeFile, mkdirp} from 'fs-extra';
```
Are we able to leverage the [`FileSystem` class](https://github.com/lemonmade/sewing-kit/blob/main/packages/model/src/base/fs.ts) instead? It should be accessible through `pkg.fs`.

I haven't confirmed this yet with @lemonmade, but I believe the intention is to always use the `FileSystem` class whenever possible. I could be wrong here but I believe its so we can have a better e2e testing solution. We'll just have to mock the `Filesystem` class instead of a virtual filesystem.

b-ryu replied:
I think in the case of `pkg.fs` it's scoped to `pkg.root` (e.g. `<workspace_root>/packages/plugin-javascript`), which makes it difficult to do cache-related stuff in `.sewing-kit/cache`.

That said I could just create a new `FileSystem` instance for the cache 👍

ismail-syed commented on packages/plugin-javascript/src/utilities.ts:
```diff
@@ -224,6 +256,81 @@ export function createCompileBabelStep({
+  );
+}
+
+function generateBabelPackageCacheValue(
```
Would love to see tests for this and `getLatestModifiedTime`.

ismail-syed replied:
There's also a lot of parameters in there. Can we break this up a bit?

eg.
```
interface BabelCacheOptions { 
  babelConfig: BabelConfig,
  outputPath: string,
  extension: string,
  exportStyle: ExportStyle,
  babelCacheDependencies: string[],
  babelIgnorePatterns: string[],
  babelExtensions: string[],
}

function generateBabelPackageCacheValue(pkg: Package, options: BabelCacheOptions)...

```

ismail-syed commented on packages/plugin-javascript/src/utilities.ts:
```diff
@@ -166,8 +170,36 @@ export function createCompileBabelStep({
+          pkg,
+          babelConfig,
+          outputPath,
+          extension || '.js',
```
Is there anything preventing us from setting these defaults (`extension` and `exportStyle`) in the function params? Having all the defaults in one place is easier to read. Similar to `cacheDependencies` defaulting to `[]` in the function params.

ismail-syed reviewed (changes requested):
Left a few comments. Looking good so far, aside from adding some test coverage. Can you also share some 🎩  instructions? Maybe we can morphe them into 🎩  docs for the codebase.

b-ryu commented:
> Left a few comments. Looking good so far, aside from adding some test coverage. Can you also share some 🎩 instructions? Maybe we can morphe them into 🎩 docs for the codebase.

Sure thing. I tested the new changes in two places mostly, SK-next itself and Quilt.

**SK-next**
Once you've pulled down this branch it should just be a matter of `yarn build`, since SK-next builds itself. Running a second `yarn build` should take less time (about five seconds faster than before), and should print out some relevant logs.

**Quilt**
Quilt was a little trickier, since I ran into some problems `yarn link`ing local SK-next packages into Quilt. I ended up just copy-pasting built SK-next packages into `quilt/node_modules` using a script 😅(I added a note about tophatting local SK-next to the migration doc issue, probably tools like `yalc` might make things easier).

Make sure that you're on the [Quilt branch that uses SK-next](https://github.com/Shopify/quilt/tree/1522-sk-next), and that you've `yarn build`ed in the SK-next branch. Then paste this into a file and run it:

```js
// Basically copies build directories right into node_modules

const {copySync} = require('fs-extra');
const rimraf = require('rimraf');

// Change these consts to match your file system
const PATH_TO_QUILT_NODE_MODULES =
  '<path_to_quilt>/Shopify/quilt/node_modules';
const PATH_TO_LOCAL_SEWING_KIT =
  '<path_to_sk_next>/b-ryu/sewing-kit';

// Packages I changed in the PR
const packages = [
  'plugin-javascript',
  'cli',
  'plugin-package-node',
  'plugin-package-commonjs',
  'plugin-package-esmodules',
  'plugin-package-esnext',
];

packages.forEach(package => {
  const nodeModulePath = `${PATH_TO_QUILT_NODE_MODULES}/@sewing-kit/${package}/build`;
  const buildPath = `${PATH_TO_LOCAL_SEWING_KIT}/packages/${package}/build`;

  rimraf.sync(nodeModulePath);
  copySync(buildPath, nodeModulePath);
});
```

After that, run `yarn build` as in SK-next. The second, warm build should take less than 10s.

Commit:
Extract params

Commit:
Extract defaults

Commit:
Use FileSystem class

Commit:
Fix typo in test

Commit:
Add tests for utilities

Commit:
Add tests for Babel package caching

Commit:
Use functions directly in tests

ismail-syed commented on packages/plugin-javascript/src/utilities.ts:
```diff
@@ -224,11 +257,101 @@ export function createCompileBabelStep({
+  pkg: Package,
+  options: BabelPackageCacheOptions,
+) {
+  const {
+    babelConfig,
+    outputPath,
+    extension,
+    exportStyle,
+    babelCacheDependencies,
+    babelIgnorePatterns,
+    babelExtensions,
+  } = options;
```
Super ultra nit 😅 : you can pull these out in the function params. 

```tsx
export function generateBabelPackageCacheValue(
  pkg: Package,
  {babelConfig, outputPath.... }: BabelPackageCacheOptions,
)
```

ismail-syed commented on packages/plugin-javascript/tests/plugin-javascript.test.ts:
```diff
@@ -0,0 +1,137 @@
+      it('creates a cache', async () => {
+        await withWorkspace('simple-package', async (workspace) => {
+          await workspace.writeConfig(babelCompilationConfig);
+
```
🔥 . We don't need this line. IMO, we can group all the _write_ related things. 
```
          await workspace.writeConfig(babelCompilationConfig);
          await writeToSrc(workspace, 'index.js');

          await workspace.run('build');
```

ismail-syed commented on packages/plugin-javascript/tests/plugin-javascript.test.ts:
```diff
@@ -0,0 +1,137 @@
+          await workspace.run('build');
+
+          const builtOutputModifiedTime = getModifiedTime(
+            resolve(workspace.root, 'build', 'esm', 'index.mjs'),
```
Let's pull this path out into the top of each test and re-use it instead.

ismail-syed commented on packages/plugin-javascript/tests/utilities.test.ts:
```diff
@@ -0,0 +1,351 @@
+      });
+    });
+
+    it('excludes files with extensions not included in the provided list', async () => {
```
I could be missing something here. But how does this test description line up with the assertion?

ismail-syed replied:
Oh, is it because this test is only building `.js` files and it's adding a `.ts` file to the workspace which shouldn't impact the build.

ismail-syed replied:
Let's re-phrase the test description a bit then to make it easier for readers. 

Maybe: 
```
excludes .ts file extensions when it is not accounted for in fileExtensions 
```

ismail-syed commented on packages/plugin-javascript/tests/utilities.test.ts:
```diff
@@ -0,0 +1,351 @@
+
+        const newOptions = {
+          ...options,
+          outputPath: 'build/meme',
```
😄

ismail-syed commented on packages/plugin-javascript/tests/utilities.test.ts:
```diff
@@ -0,0 +1,351 @@
+      });
+    });
+
+    it('generates a different hash if the Babel config changes', async () => {
```
There's a lot of duplication across these `generates a different hash if .... changes` tests. Can we reduce all this bloat?

Some ideas: 

* Create a module-level utility so you can do a one-liner for `const testPackage = createTestPackage('simple-package, workspace.root)`
* Do we need the `writeToSrc` calls at all since we just care about the options passed in? If we need 'em lets stick to just one call. Unless I'm missing something here and we need all 3.


Something like, perhaps? 

```tsx
      await withWorkspace('simple-package', async (workspace) => {
        const testPackage = createTestPackage('some-package', workspace.root);
       // add writeToSrc calls if we need 'em

        const newOptions = {
          ...options,
          babelConfig: {
            ...options.babelConfig,
            plugins: [...options.babelConfig.plugins, '@babel/some-new-plugin'],
          },
        };

        expect(
            generateBabelPackageCacheValue(testPackage, options)
         ).not.toEqual(
            generateBabelPackageCacheValue(testPackage, newOptions)
         );
      });
```

ismail-syed commented:
Did a 🎩  against sk-next itself. Getting the same perf improvements you noted. Went from a cold build of `~192s` to a warm build of `~8s`. 

I also did another 🎩  by added a `console.log` in the `plugin-webpack` and verifying that it does rebuild only that package and it did as expected. This build took `32.85s`. Still some room for improvement there but we can always revisit this. I'm more than happy with the improvements in this PR. 

@themallen, do you also mind giving this some 👀 .

Commit:
Address feedback

ismail-syed commented:
Moved to https://github.com/Shopify/sewing-kit-next/pull/4
